File: src/atforge/api/schemas/runs.py
"""Pydantic models for /runs endpoints."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel


class RunSummary(BaseModel):
    run_id: str
    started_at: datetime | None
    finished_at: datetime | None
    status: Literal["running", "done", "failed", "unknown"]
    n_symbols: int | None
    max_generations: int | None
    current_generation: int | None
    n_backtests: int
    n_failures: int


class RunListResponse(BaseModel):
    runs: list[RunSummary]
    total: int
    limit: int
    offset: int


class EventEnvelope(BaseModel):
    event_id: int
    run_id: str
    ts_ms: int
    event_type: str
    generation: int | None
    payload: dict[str, Any]


class RankingRow(BaseModel):
    backtest_id: int
    symbol: str
    strategy_id: int
    strategy_name: str
    family: str
    generation: int
    n_trades: int
    sharpe: float | None
    sortino: float | None
    cagr: float | None
    win_rate: float | None
    max_drawdown: str | None
    total_return: str | None


class RunRankingsResponse(BaseModel):
    run_id: str
    rankings: list[RankingRow]


class ExperimentRow(BaseModel):
    experiment_id: int
    generation: int
    mutator: str | None
    accepted: int | None
    delta_sharpe: float | None
    reasoning: str | None
    composite_score: str | None
    mutation_json: str | None
    created_at: str | None
    parent_name: str | None
    child_name: str | None
    parent_strategy_id: int | None
    child_strategy_id: int | None


class GenerationSharpe(BaseModel):
    generation: int
    best_sharpe: float | None
    n_backtests: int


class RunEvolutionResponse(BaseModel):
    run_id: str
    experiments: list[ExperimentRow]
    sharpe_progression: list[GenerationSharpe]


class TimelineResponse(BaseModel):
    run_id: str
    events: list[EventEnvelope]

File: web/src/lib/api/runs.ts
import { apiGet } from "./client";
import type { components } from "./types.gen";

export type RunSummary = components["schemas"]["RunSummary"];
export type RunListResponse = components["schemas"]["RunListResponse"];
export type EventEnvelope = components["schemas"]["EventEnvelope"];
export type RunRankingsResponse = components["schemas"]["RunRankingsResponse"];
export type RankingRow = components["schemas"]["RankingRow"];
export type RunEvolutionResponse = components["schemas"]["RunEvolutionResponse"];
export type ExperimentRow = components["schemas"]["ExperimentRow"];
export type TimelineResponse = components["schemas"]["TimelineResponse"];

export async function fetchRuns(limit = 20, offset = 0): Promise<RunListResponse> {
  return apiGet<RunListResponse>(`/runs?limit=${limit}&offset=${offset}`);
}

export async function fetchRun(runId: string): Promise<RunSummary> {
  return apiGet<RunSummary>(`/runs/${encodeURIComponent(runId)}`);
}

export async function fetchRunRankings(runId: string): Promise<RunRankingsResponse> {
  return apiGet<RunRankingsResponse>(`/runs/${encodeURIComponent(runId)}/rankings`);
}

export async function fetchRunEvolution(runId: string): Promise<RunEvolutionResponse> {
  return apiGet<RunEvolutionResponse>(`/runs/${encodeURIComponent(runId)}/evolution`);
}

export async function fetchRunTimeline(runId: string): Promise<TimelineResponse> {
  return apiGet<TimelineResponse>(`/runs/${encodeURIComponent(runId)}/timeline`);
}

export const runsQueryKeys = {
  all: ["runs"] as const,
  detail: (runId: string) => ["runs", runId] as const,
  rankings: (runId: string) => ["runs", runId, "rankings"] as const,
  evolution: (runId: string) => ["runs", runId, "evolution"] as const,
  timeline: (runId: string) => ["runs", runId, "timeline"] as const,
};
